src/libs/actions/Report.js

```diff
@@ -1017,8 +964,17 @@ function deleteReportComment(reportID, reportAction) {
         ],
     };
 
-    Onyx.merge(`${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${reportID}`, reportActionsToMerge).then(() => {
-        setLocalLastRead(reportID, getLastReadSequenceNumber(reportID));
+    // If the comment we are deleting is more recent than our last read comment we will update the unread count
+    if (sequenceNumber > lodashGet(allReports, [reportID, 'lastReadSequenceNumber'])) {
+        Onyx.merge(`${ONYXKEYS.COLLECTION.REPORT}${reportID}`, {
+            unreadActionCount: Math.max(lodashGet(allReports, [reportID, 'unreadActionCount'], 0) - 1, 0),
+        });
+    }
+
+    // Optimistically update the report and reportActions
+    Onyx.merge(`${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${reportID}`, reportActionsToMerge);
+    Onyx.merge(`${ONYXKEYS.COLLECTION.REPORT}${reportID}`, {
+        lastMessageText: ReportActions.getLastVisibleMessageText(reportID, reportActionsToMerge),
```

Comment:
Great question. So, we were doing this before, but the way it happens is:

- current user deletes comment
- a Pusher event is sent out and this code is called:
https://github.com/Expensify/App/blob/0fcc4920dbc0e0b7658e80541367acf8a2e4ff57/src/libs/actions/Report.js#L562
- `updateReportActionMessage()` then does this:
https://github.com/Expensify/App/blob/0fcc4920dbc0e0b7658e80541367acf8a2e4ff57/src/libs/actions/Report.js#L523-L532

There are a few issues with this that came to mind...

- `Onyx.merge().then()` feels like an anti-pattern because we can use the new action and existing actions to determine the last message
- It should be updated optimistically when we delete the comment and not only when a Pusher event is handled (which means it was successful)

Since this event will still be sent to the person who deleted the comment I added a check so that we will not decrement if they are the one deleting the comment.

Reply:
That makes sense, thanks for explaining.